File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/api/util/EventUtil.kt
package io.nativeblocks.runtime.api.util

import io.nativeblocks.runtime.api.provider.block.BlockContext
import io.nativeblocks.runtime.api.provider.modifier.ModifierContext
import io.nativeblocks.runtime.api.provider.model.NativeActionModel

private fun blockHandleEvent(blockContext: BlockContext, action: NativeActionModel, eventType: String) {
    blockContext.onHandleAction.invoke(
        blockContext.listItemIndex, action, eventType
    )
}

fun blockProvideEvent(
    blockContext: BlockContext,
    eventType: String
): (() -> Unit)? {
    val action = blockContext.onFindAction.invoke(eventType)
    return if (action != null) {
        { blockHandleEvent(blockContext, action, eventType) }
    } else {
        null
    }
}

private fun modifierHandleEvent(
    modifierContext: ModifierContext,
    action: NativeActionModel,
    eventType: String
) {
    modifierContext.onHandleAction.invoke(
        modifierContext.listItemIndex, action, eventType
    )
}

fun modifierProvideEvent(
    modifierContext: ModifierContext,
    eventType: String
): (() -> Unit)? {
    val action = modifierContext.onFindAction.invoke(eventType)
    return if (action != null) {
        { modifierHandleEvent(modifierContext, action, eventType) }
    } else {
        null
    }
}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/api/util/EventUtil.swift
import Foundation

/// Provides an event handler for a specific event type if the block supports it.
///
/// - Parameters:
///   - blockContext: The properties of the block, including its event handling logic.
///   - eventType: The type of event to check and provide a handler for.
/// - Returns: A closure representing the event handler if the event type is supported; otherwise, `nil`.
public func blockProvideEvent(blockContext: BlockContext, eventType: String) -> (() -> Void)? {
    if let action = blockContext.onFindAction(eventType) {
        return { blockHandleEvent(blockContext: blockContext, action: action, eventType: eventType) }
    } else {
        return nil
    }
}

/// Handles an event for a block by invoking the appropriate action handler.
///
/// - Parameters:
///   - blockContext: The properties of the block, including its event handling logic.
///   - action: The action associated with the block for this event.
///   - eventType: The type of event to handle.
private func blockHandleEvent(blockContext: BlockContext, action: NativeActionModel, eventType: String) {
    blockContext.onHandleAction(blockContext.listItemIndex, action, eventType)
}
